icu4c/source/common/rbbi_cache.cpp

```diff
@@ -420,50 +422,128 @@ UBool RuleBasedBreakIterator::BreakCache::populateNear(int32_t position, UErrorC
     return true;
 }
 
+static inline bool isNumber(UChar32 ch) {
+    static UChar32 range[2][2] = {
+        {0x0030, 0x0039},    //half width number 0~9
+        {0xFF10, 0xFF19}     //full width number 0~9
+    };
+    for(int32_t i = 0 ; i < 2 ; i++) {
+        if(ch >= range[i][0] && ch <= range[i][1])
+            return true;
+    }
+    return false;
+}
+
+static inline bool isCjk(UChar32 ch) {
```

Comment:
If possible, this would be better expressed as a UnicodeSet, using Unicode properties where possible.

Reply:
About Mark's comment, take a look at   UnicodeSet    fSet; in DictionaryBreakEngine :
and how is it used in 
```
UBool
DictionaryBreakEngine::handles(UChar32 c) const {
    return fSet.contains(c);
}
```

The code which construct them now in each LanguageBreakEngine seems has some problem (I see no point to keep them in the member data and please ignore those.

Reply:
I found some of the UnicodeSet usage is wrong when I try to look at example for you. Try to correct them in https://github.com/unicode-org/icu/pull/1941

Reply:
```
UnicodeSet cjkCharacters(UNICODE_STRING_SIMPLE("[[:Han:][:Hiragana:][:Hangul:][:Katakana:]]"), status);
```
should match all the Han, Hiragana, Katakana, Hangul letters, if you want to add more punchation to this list you may need to add some more to the set. 

I am not sure they have [:Jamo:] or not . read the doc to figure out what should you put into the set.
See
https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/classicu_1_1UnicodeSet.html

Reply:
read https://unicode-org.github.io/icu/userguide/strings/unicodeset.html
@markusicu  - is there a clear documentation to list Hiragana, Katakana and other acceptable id used inside UnicodeSet somewhere. I saw a lot of indirect link but it is rather hard to find the table which list all the acceptable keys. For example, how can I tell [:Jamo:] or [:FooBar:] is an acceptable expression or not from the documentation ?

Reply:
Please help review again. I integrated #1941 and used DictionaryBreakEngine#handles() to implement isCjk()